prime-router/docs/design/validation/design.md

```diff
@@ -0,0 +1,284 @@
+- Behavior around how the parser should encode segments and fields even if no data content is present
+
+The HAPI HL7 library provides the following classes to validate HL7 messages:
+- `ValidationContext` - manages a collection of codified rules for handling messages, fields, and primitive types
+- `ValidationRuleBuilder` - what you use to create the validation rules
+- `MessageVisitor` - the interface the classes implement when they are in charge of doing the actual comparison
+- `ValidationException` - thrown when validation fails
+- `ValidationExceptionHandlerFactory` - interface that you can implement in order to control the exception handling
+
+Code Examples:
+
+A code example for custom message validation: 
+https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/CustomMessageValidation.html  
+If you want to start with something more basic for a POC, here is an example that uses a default validation context
+rather than building one from scratch: https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/MessageValidation.html   
+Here is an example of using the ValidationRuleBuilder: 
+https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/CustomMessageValidationWithValidatingVisitor.html   
+An example for using conformance: 
+https://hapifhir.github.io/hapi-hl7v2/xref/ca/uhn/hl7v2/examples/MessageValidationUsingConformanceProfile.html   
+
+##### NIST Validator
+The NIST HL7 Validation Library can be found here: https://github.com/usnistgov/v2-validation  
+Unfortunately, they do not seem to have documentation on how to actually use this library. I found this page
+https://hl7v2-ws.nist.gov/hl7v2ws/documentation.htm, but the buttons leading to the documentation do not work. The 
+project only has one README which does not discuss how to actually use the library. https://hl7v2-gvt.nist.gov/gvt/#/doc 
+appears to have information on the validation tool and within that there is a swagger spec
+(https://hl7v2-gvt.nist.gov/gvt/apidocs/swagger-ui.html#/), but none of that appears to be for actually validating an
+HL7 Message. 
+
+DEX uses this library but has a light wrapper around it. They do recommend using the library, but not their wrapper.
+Part of using the library is creating profiles. This can be done by going to https://hl7v2-igamt-2.nist.gov/ and signing
+up. Once you register, go to IG Documents, Create new IG document. Select the HL7 version as well as the message structure and 
+it will create a base for you. Then you have to manually adjust each field to meet the spec. There is only one profile 
+currently that is shared. Otherwise, you have to know someone. Marcia Schulman has offered to give us whatever she has 
+so that she can be contacted once we know what profiles we need, but she only had a few and it is unlikely they will 
+meet our needs. I discussed using their API with them, but they said that we cannot send PII through the API, so this 
```

Comment:
`I discussed using their API with them.` This  and the rest of the sentence should be rewritten to be more clear: "NIST hosts a validation API based on the NIST library but it cannot be used for production data due to concerns over PII."